oxidation/libparsec/crates/serialization_format/src/protocol.rs

```diff
@@ -0,0 +1,739 @@
+                    major_version,
+                    cmd.req.cmd
+                );
+                gen_versions.entry(major_version).or_default().push(gen_cmd);
+            }
+        }
+
+        Self {
+            name: familly_name.to_owned(),
+            versions: gen_versions,
+        }
+    }
+}
+
+//
+// Code generation
+//
+
+impl GenCmdsFamilly {
+    pub(crate) fn quote(&self) -> TokenStream {
+        let familly_name = format_ident!("{}", &self.name);
+
+        let versioned_cmds: Vec<TokenStream> = self
+            .versions
+            .iter()
+            .sorted_by_key(|(v, _)| *v)
+            .map(|(version, cmds)| quote_versioned_cmds(*version, cmds))
+            .collect();
+
+        quote! {
+            pub mod #familly_name {
+                #(#versioned_cmds)*
+            }
+        }
+    }
+}
+
+fn quote_versioned_cmds(version: u32, cmds: &[GenCmd]) -> TokenStream {
+    let versioned_cmds_mod = format_ident!("v{version}");
+    let (any_cmd_req_variants, cmd_structs): (Vec<TokenStream>, Vec<TokenStream>) =
+        cmds.iter().map(quote_cmd).unzip();
+
+    quote! {
+        pub mod #versioned_cmds_mod {
+            // Define `UnknownStatus` here instead of where is it actually used (i.e.
+            // near each command's `Rep::load` definition) to have a single common
+            // definition that will have it deserialization code compiled once \o/
+            #[derive(::serde::Deserialize)]
+            struct UnknownStatus {
+                status: String,
+                reason: Option<String>
+            }
+
+            #[derive(Debug, Clone, ::serde::Serialize, ::serde::Deserialize, PartialEq, Eq)]
+            #[serde(tag = "cmd")]
+            pub enum AnyCmdReq {
+                #(#any_cmd_req_variants),*
+            }
+
+            impl AnyCmdReq {
+                pub fn dump(&self) -> Result<Vec<u8>, ::rmp_serde::encode::Error> {
+                    ::rmp_serde::to_vec_named(self)
+                }
+
+                pub fn load(buf: &[u8]) -> Result<Self, ::rmp_serde::decode::Error> {
+                    ::rmp_serde::from_slice(buf)
+                }
+            }
+
+            #(#cmd_structs)*
+        }
+    }
+}
+
+fn quote_cmd(cmd: &GenCmd) -> (TokenStream, TokenStream) {
+    let camel_case_name = &snake_to_camel_case(&cmd.cmd);
+    let snake_case_name = &cmd.cmd;
+
+    let variant_name = format_ident!("{}", camel_case_name);
+    let module_name = format_ident!("{}", snake_case_name);
+    let command_name = snake_case_name;
+
+    let module = cmd.quote();
+    let variant = quote! {
+        #[serde(rename = #command_name)]
+        #variant_name(#module_name::Req)
+    };
+
+    (variant, module)
+}
+
+impl GenCmd {
+    pub(crate) fn quote(&self) -> TokenStream {
+        let variant_name = format_ident!("{}", &snake_to_camel_case(&self.cmd));
+        let module_name = format_ident!("{}", &self.cmd);
```

Comment:
Could you provide those values as argument ? 
That way we're sure to use the same value as in `quote_cmd` since it also create those values

Reply:
Even better: merge `GenCmd::quote` into `quote_cmd` !

I don't know why I put a `GenCmd::quote`: having structs without methods is good given we divide here the code between 1) json structs 2) gen structs 3) json->gen conversion 4) code generation , so having each json/gen struct implementing it own subpart of the code would make much more spaghetti code

I've also replaced `GenCmdsFamilly::quote` by a `quote_cmds_familly` to stay consistent.

As a matter of fact I also pondering removing `GenCmdsFamilly::new`, but given it already contains all the json->gen conversion, turning it into a function would only gain removing a level of indentation.


Aaand finally I've removed the `pub(crate)` on most of the structure: external module only needs to know about `JsonCmd` (need given caller code is responsible for reading the json files and hence can deserialize each json file independently for better error reporting) and `generate_protocol_cmds_familly` (which wrap `GenCmdsFamilly::new/GenCmdsFamilly::quote`).